packages/squiggle-lang/src/rescript/FR/FR_Danger_Aggregate.res

```diff
@@ -0,0 +1,171 @@
+            | (_, false, false) => Ok(x)
+            }
+          , xs)
+          let checkedCollectiveXs = E.A.R.firstErrorOrOpen(checkedIndividualXs)
+          let result = switch checkedCollectiveXs {
+          | Ok(xs) => fn(xs)
+          | Error(e) => Error(e)
+          }
+          result
+        }
+      }
+    }
+    let probabilityToOdds = p => p /. (1.0 -. p)
+    let oddsToProbability = o => o /. (1.0 +. o)
+    let sum = xs => E.A.reduce(xs, 0.0, (a, b) => a +. b)
+    let mean = xs => sum(xs) /. Belt.Float.fromInt(E.A.length(xs))
+
+    let geomMean = (xs: array<float>) => {
+      let xsLogs = E.A.fmap(x => Js.Math.log2(x), xs)
+      let meanXsLogs = mean(xsLogs)
+      let answer = Js.Math.pow_float(~base=2.0, ~exp=meanXsLogs)
+      answer
+    }
+    let arithmeticMean = (xs: array<float>) => {
+      let sumXs = E.A.reduce(xs, 0.0, (a, b) => a +. b)
+      let meanXs = sumXs /. Belt.Float.fromInt(E.A.length(xs))
+      meanXs
+    }
+
+    let geomMeanOfOdds = xs => {
+      let arrayOfOdds = E.A.fmap(p => probabilityToOdds(p), xs)
+      let arrayOfLogsOfOdds = E.A.fmap(p => Js.Math.log2(p), arrayOfOdds)
+      let meanOfLogsOfodds = mean(arrayOfLogsOfOdds)
+      let geomMeanOfOdds = Js.Math.pow_float(~base=2.0, ~exp=meanOfLogsOfodds)
+      let result = oddsToProbability(geomMeanOfOdds)
+      result
+    }
+
+    let extremizedGeometricMeanOfOdds = (~xs, ~extremizationParameter=1.5, ()) => {
+      let arrayOfOdds = E.A.fmap(p => probabilityToOdds(p), xs)
+      let arrayOfLogsOfOdds = E.A.fmap(p => Js.Math.log2(p), arrayOfOdds)
+      let extremizedSumOfLogsOfOdds = extremizationParameter *. mean(arrayOfLogsOfOdds)
+      let extremizedGeomMeanOfOdds = Js.Math.pow_float(~base=2.0, ~exp=extremizedSumOfLogsOfOdds)
+      let result = oddsToProbability(extremizedGeomMeanOfOdds)
+      result
+    }
+
+    let neyman = xs => {
+      let n = Belt.Int.toFloat(E.A.length(xs))
+      let d =
+        n *.
+        (Js.Math.sqrt(3.0 *. Js.Math.pow_float(~base=n, ~exp=2.0) -. 3.0 *. n +. 1.0) -. 2.0) /.
+        (Js.Math.pow_float(~base=n, ~exp=2.0) -. n -. 1.0)
+      let result = extremizedGeometricMeanOfOdds(~xs, ~extremizationParameter=d, ())
+      result
+    }
+
+    let samotsvety = xs => {
+      let sortedXs = Belt.SortArray.stableSortBy(xs, (a, b) => a > b ? 1 : -1)
+      let middleXs = Belt.Array.slice(sortedXs, ~offset=1, ~len=E.A.length(xs) - 2)
+      Js.log2("MiddleXs: ", middleXs)
+      let answer = geomMeanOfOdds(middleXs)
+      answer
+    }
+  }
+  module Lib = {
+    let geomMean = Function.make(
+      ~name="geomMean",
+      ~nameSpace,
+      ~output=EvtNumber,
+      ~requiresNamespace=true,
+      ~examples=[`Danger_Aggregate.geomMean([0.1, 0.2, 0.4])`],
+      ~definitions=[
+        DefineFn.Numbers.arrayToNum("geomMean", xs =>
+          Helpers.checker(xs => Ok(Helpers.geomMean(xs)), xs, 2)
+        ),
+      ],
+      (),
+    )
```

Comment:
we already have mean and geoMean in the API. https://www.squiggle-language.com/docs/Api/Number#geometric-mean

Reply:
I think I'd want to make the call to have them both as functions and as aggregation functions. In particular, I am imagining a use case where I am applying many aggregation functions at once.

```
xs = [...]
[Aggregate.method1(xs), Aggregate.method2(xs), ....]

```

In that case, I want geomean to also be in the Aggregate namespace, for consistency.

Reply:
Yea, that makes sense to me. This isn't something we should do often, but in this case it seems reasonable.

Reply:
Also the previous geommean algorithm was suboptimal, changed now.